### ISSUE TITLE
ProcessWire 2.7.1: search function returns nothing when search string contains "®" + following char(s)

### ISSUE THREAD

**Issue by processwired** [ryancramerdesign/ProcessWire#1542]
I have some of pages which use the trademark sign "®" in their titles, for example: "test® test".

If I search for "test®", the results are found.
If I search for "test® " (one blank after the ®), the results are found.
If I search for "test® t", no result is found (?)

It seems that ProcessWire does not find anything if a character comes after ®, with the exception of a space (I guess that spaces are trimmed anyway).

I suppose it has to do with how $sanitizer->selectorValue() works.

**Comment by teppokoivula** [comment]
Which selector operator are you using?

**Comment by processwired** [comment]
$query = $sanitizer->selectorValue('test® test');
$res = $pages->find("title%=$query");

**Comment by teppokoivula** [comment]
This is actually [documented behaviour](https://github.com/ryancramerdesign/ProcessWire/blob/dev/wire/core/Sanitizer.php#L799) of selectorValue(): "replaces non-alphanumeric and non-space with space". The way I would interpret this is that running "test® test" through selectorValue() *should* indeed it to "test test", as it does (and that explains the results you're getting here).

There's definitely weird stuff going on with selectorValue(): it doesn't remove *all* non-alphanumeric characters, just some of them. Even when it correctly recognises that the selector string contains non-alphanumeric characters, it still doesn't necessarily remove them, which is downright weird.

On the other hand, if it did remove them (like the comment states), that would in my opinion be a problem. Perhaps it's just me, but I wasn't aware of this behaviour, and wouldn't have expected selectorValue() to remove characters like "+", even if it's not alphanumeric, and thus should be removed.

The comments and actual behaviour are not in line here, so @ryancramerdesign needs to verify which is the expected behaviour. Depending on that we can say for sure if removing ® is to be expected :)

**Comment by ryancramerdesign** [comment]
The comment uses alphanumeric in the broadest possible definition. It really means it removes non-word characters. I can update that comment. Overall the selectorValue function might need to be loosened up too.

**Comment by teppokoivula** [comment]
Thanks, Ryan.

In addition to that particular comment, the use of PHP's ctype_alnum() feels a bit weird here – especially since if the string isn't alphanumeric based on that test, another comment declares that it "needs more filtering", yet the following preg_replace and str_replace calls won't necessarily change anything at all.

Obviously there's not really much of a performance penalty in that, but it does make interpreting the real purpose of said method difficult :)

**Comment by ryancramerdesign** [comment]
Teppo, note that ctype_alnum() has an exclamation point before it, meaning NOT ctype_alnum. The purpose of that is just to avoid executing slower preg_replace functions if we don't need to.

**Comment by teppokoivula** [comment]
@ryancramerdesign: I do know how negation works in PHP :)

My point was that if the string enters that if block, the next comment states that it "needs filtering", which often isn't true at all. Only thing worse than no comments are comments that are not in sync with actual behaviour.

**Comment by ryancramerdesign** [comment]
I know you know what negation is, but your previous comment made me think maybe you didn't see the exclamation point preceding the ctype_alnum call, so was just trying to clarify. 

Regardless, I guess I'm not following what you mean. The comment describes to me at least exactly what it is doing. Though you've got me concerned there's some obvious bug in here that I'm missing. Or are you just saying that it should say "potentially needs more filtering" rather than "needs more filtering"? We're looking at code here, and if something *potentially* needs more filtering, then it needs more filtering – that's a given, no? I'm not sure any of my code comments can live up to your standards. :) Though if this is really what you mean, then I'm happy to adjust the wording there. Chances are I'm missing something obvious and just don't understand what you are talking about, so I'll take another look after my 2nd coffee has kicked in.

**Comment by teppokoivula** [comment]
It's probably just a matter of different views:

* In my opinion if you state that "this string needs filtering", it must really need it; i.e. it must contain something invalid.

So yes, it's mostly a question of wording here. If the string is already valid, it *doesn't* need any more filtering, and thus the claim that it does is not true. It's as simple as that. The wording matters to me only because here I was trying to interpret how the method should work, and I couldn't; it says one thing and does another.

* The thing that confused me about !ctype_alnum() is that a string may enter the if block even if it doesn't require any filtering at all.

I get your point about not wanting to use regex when it can be avoided, but sometimes such optimisations make code more difficult to interpret: "the if clause returned true, so my string seems to contain invalid characters.. but why doesn't the following str/preg_replace remove those, is that not working properly?"

In this case it seems that our interpretations of "non-alphanumeric" are quite different. In my opinion "+", for an example, is not alphanumeric. To me alphanumeric means [A-Za-z0-9]. PHP agrees with me here, and to my best knowledge that's the most common interpretation overall.

* There are no "bugs" here as far as I can tell, but the method doesn't work in a logical way either.

First there's the mismatch between comments and actual behaviour, and on the other hand if the method is (as the name suggests) supposed to validate a selector string, why would it strip characters that are perfectly valid in selectors? This is what happened here, and it has caught me by surprise before too.

**Comment by ryancramerdesign** [comment]
Per my first comment, I'm using the term "alphanumeric" broadly to refer to word characters and separators. It's not really the correct use of the term I know, and if it's a concern I'll update that terminology when I'm working in the sanitizer next. 

As for whether or not something needs filtering, we're talking about sanitizers and validators here. Throughout all the methods in that class, there is an underlying assumption that everything is dirty until proven clean. Meaning everything needs filtering until it's proven not to. The ctype_alnum() is there to prove it as not needing filtering. At that point in the code, if the ctype_alnum() returns false, then the string has not yet been proven to be clean, and thus needs more filtering. So I stand by the use of comments here because I think they are totally correct. If you are getting caught up with that comment, then you are going to get caught up in nearly every comment in the class because everything there is coded from the same assumption of "dirty until proven clean", and this is how I think it should be in the context of a class doing sanitization. However, if it's a big concern and you want to propose a different wording I'm happy to change it too.

**Comment by teppokoivula** [comment]
I would like to suggest some changes, but to be fair: I'm not going to argue that my interpretation is the only right one out there, and I get your point about assuming that everything is dirty. I prefer to do the same thing in my own code.

When it comes to comments (or code readability for that matter), whenever there's more than one way to interpret something I've written, sometimes I find myself spending ridiculous amounts of time trying to find an unambiguous way to say or do the same thing. If everyone was the same, we probably wouldn't have systems the size of ProcessWire, so hats off to you for focusing on what matters the most :)

Regardless, the original issue is still unsolved: selectorValue() prevents using certain characters in a selector value, even if they're otherwise OK. Would you consider "loosening" the current method a bit to accommodate such cases?

**Comment by ryancramerdesign** [comment]
Definitely, I agree it needs to be loosened up a bit. And I've been doing this over time. The first iteration back in PW 2.0 literally was alphanumeric and space, not so multi-language friendly! But I've been loosening it up in small steps just to stay as security mindful of it was possible. What is the range of characters you think we should loosen it up to support? Just the "+", "®", and "©" or are there others you've run into that you'd like it to support?

**Comment by processwired** [comment]
I'm fine with that, especially with "®" ;-)
